src/net/delta.cpp

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (c) 2020 The reone project contributors
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#include <cstdint>
+#include <array>
+#include <vector>
+#include <tuple>
+
+#include "delta.h"
+#include "../mp/objects.h"
+
+using namespace std;
+
+namespace reone {
+
+namespace net {
+
+template <typename ... Args, size_t ... Is>
+auto getSerImpl(tuple<Args*...> &&seqPtr, integer_sequence<size_t, Is...>) {
+    return make_tuple(*(get<Is>(seqPtr))...);
+}
+
+template <typename T>
+auto getSerial(T &obj) {
+    return getSerImpl(getSerPtr(obj), make_index_sequence<tuple_size<decltype(getSerPtr(obj))>::value>());
+}
+
+/* helper functions */
+
+using expander = int[]; // https://stackoverflow.com/a/25683817
+
+template <size_t I, typename ... Args>
+void diffHeaderIter(const tuple<Args...> &left, const tuple<Args...> &right, uint16_t &header) {
+    if (get<I>(left) != get<I>(right)) {
+        header |= 1 << I;
+    }
+}
+
+template <size_t ... Is, typename ... Args>
+uint16_t diffHeader(const tuple<Args...> &left, const tuple<Args...> &right, integer_sequence<size_t, Is...>) {
```

Comment:
Can we give `tuple<Args...>` a type name, e.g. `typedef tuple<T...> SerialMemberPointers<T>`?